hazelcast/src/main/java/com/hazelcast/collection/impl/collection/CollectionContainer.java

```diff
@@ -335,6 +344,39 @@ public void destroy() {
 
     protected abstract void onDestroy();
 
+    @Override
+    public CollectionItem merge(SplitBrainMergeEntryView<Long, Data> mergingEntry, SplitBrainMergePolicy mergePolicy) {
+        mergePolicy.setSerializationService(nodeEngine.getSerializationService());
+
+        // try to find an existing item with the same value
+        CollectionItem existingItem = null;
```

Comment:
I know it's not a clear cut decision but I'm thinking if we should find the existing entry by value or by ID. E.g. how should we behave if we are merging these two lists: `[1,1,1,2,3,1]` and `[2,1,1,3,2,1]`. Right now all of the `1`s in the list from the smaller cluster will be merged with the same entry in the bigger cluster. Maybe switching to merging entries by ID would produce better results. The same comment was on ringbuffer PR AFAIK.

Reply:
The problem with the ID is, that it's completely internal and unrelated to the value. So if we add the same values to a Set in both sub-clusters, but the operations will be handled in a different order, the sets will contain the same items, but with different internal IDs. I don't think merging by those are a good idea. We should merge by what the user knows about the data structure (the contents, not the implementation details).

Reply:
Yes, there are actually good points on both approaches. The downside to merging entries with the same ID is, as you say, that the items may just be ordered differently.
The downside to merging entries that are equal are:
- in the degenerative case when the collection from the smaller cluster contains the same entry multiple times, it will be merged with the same entry in the bigger cluster. For instance, if the list in the bigger cluster is `[1,2,3,4,5]` and the list from the smaller cluster is `[1,1,1,1]`, all of entries from the smaller cluster will be merged with the same entry in the bigger cluster
- there is also the unusual case when the merge policy returns a different value which is then equal to some value further down in the merging collection. For instance, let's say there's a silly policy that returns the `value+1` and you're merging `[1,2,3,4,5]` (bigger cluster) with `[1,2,3,4,5]` (smaller cluster, merging list). First the two `1`s are merged and produce `[2,2,3,4,5]` (bigger cluster) and `[2,3,4,5]` (smaller cluster, merging collection). Now the second entry is again equal to the already merged first entry in the bigger cluster and we get `[3,2,3,4,5]` (bigger cluster) and `[3,4,5]` (smaller cluster, merging collection). Finally, after the merging collection is exhausted we end up with `[6,2,3,4,5]`.
I'm thinking that we could merge this and open an issue to consider which approach is better and if needed fix it later. Is that ok?

Reply:
In your first example nothing would actually be changed, since the values are not different after the merging. I always check `(newValue != null && !newValue.equals(oldValue))`, so if the merge policy returns the same value, we just leave everything as it is. Of course it takes some time to process everything, but nothing bad happens.

But yes, I would merge this approach (since it works in its way) and see if there are any flaws later. This applies to any keyless data structure, so we eventually have to solve it in multiple data structures anyway.